pkg/kubectl/cmd/apply_test.go

```diff
@@ -531,3 +534,89 @@ func TestApplyNULLPreservation(t *testing.T) {
 		t.Fatal("No server-side patch call detected")
 	}
 }
+
+func readTPREntryFromFile(t *testing.T, filename string) []byte {
+	raw := readBytesFromFile(t, filename)
+	obj := &unstructured.Unstructured{}
+	if err := runtime.DecodeInto(testapi.Extensions.Codec(), raw, obj); err != nil {
+		t.Fatal(err)
+	}
+
+	objJSON, err := runtime.Encode(testapi.Extensions.Codec(), obj)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return objJSON
+}
+
+// TestApplyIdempotentTPREntry check repeated apply operations on TPR entries
+// without any change.
+func TestApplyIdempotentTPREntry(t *testing.T) {
+	initTestErrorHandler(t)
+	name := "test-tpr-entry"
+	path := "/namespaces/test/testtprs/" + name
+	curr := readTPREntryFromFile(t, filenameIdempotentTPREntry)
+
+	verifiedPatch := false
+
+	f, tf, _, _ := cmdtesting.NewAPIFactory()
+	tf.Printer = &testPrinter{}
+	tf.UnstructuredClient = &fake.RESTClient{
+		APIRegistry:          api.Registry,
+		NegotiatedSerializer: unstructuredSerializer,
+		Client: fake.CreateHTTPClient(func(req *http.Request) (*http.Response, error) {
+			switch p, m := req.URL.Path, req.Method; {
+			case p == path && m == "GET":
+				body := ioutil.NopCloser(bytes.NewReader(curr))
+				return &http.Response{
+					StatusCode: 200,
+					Header:     defaultHeader(),
+					Body:       body}, nil
+			case p == path && m == "PATCH":
+				patch, err := ioutil.ReadAll(req.Body)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				patchMap := map[string]interface{}{}
+				if err := json.Unmarshal(patch, &patchMap); err != nil {
+					t.Fatal(err)
+				}
+
+				annotationsMap := walkMapPath(t, patchMap, []string{"metadata", "annotations"})
+				if _, ok := annotationsMap[annotations.LastAppliedConfigAnnotation]; !ok {
+					t.Fatalf("patch does not contain annotation:\n%s\n", patch)
+				}
+
+				verifiedPatch = true
+
+				body := ioutil.NopCloser(bytes.NewReader(curr))
+				return &http.Response{
+					StatusCode: 200,
+					Header:     defaultHeader(),
+					Body:       body}, nil
+			default:
+				t.Fatalf("unexpected request: %#v\n%#v", req.URL, req)
+				return nil, nil
+			}
+		}),
+	}
+
+	tf.Namespace = "test"
+	buf := bytes.NewBuffer([]byte{})
+	errBuf := bytes.NewBuffer([]byte{})
+
+	cmd := NewCmdApply(f, buf, errBuf)
+	cmd.Flags().Set("filename", filenameIdempotentTPREntry)
```

Comment:
Understood. 

I ran this scenario with a compiled kubectl against a running api server. 
The request of the PATCH differs in the unit test and the real example. 

In Unit Test: 
```
I0208 21:53:58.602190   21586 request.go:991] Request Body: {"metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"unit-test.test.com/v1\",\"key\":\"value\",\"kind\":\"Widget\",\"metadata\":{\"annotations\":{},\"name\":\"widget\",\"namespace\":\"test\"}}\n"}}}
I0208 21:53:58.602218   21586 request.go:991] Response Body: {"apiVersion":"unit-test.test.com/v1","key":"value","kind":"Widget","metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"unit-test.test.com/v1\",\"key\":\"value\",\"kind\":\"Widget\",\"metadata\":{\"name\":\"widget\",\"namespace\":\"test\"}}\n"},"name":"widget","namespace":"test"}}
PASS
```

In real version: 
```
I0208 21:40:17.639077    5803 request.go:991] Request Body: {"metadata":{"annotations":{}}}
I0208 21:40:17.639127    5803 round_trippers.go:398] curl -k -v -XPATCH  -H "Accept: application/json" -H "Content-Type: application/merge-patch+json" -H "User-Agent: kubectl/v0.0.0 (linux/amd64) kubernetes/$Format" https://localhost:
6443/apis/unit-test.test.com/v1/namespaces/test/widgets/widget
I0208 21:40:17.641624    5803 round_trippers.go:417] PATCH https://localhost:6443/apis/unit-test.test.com/v1/namespaces/test/widgets/widget 200 OK in 2 milliseconds
I0208 21:40:17.641641    5803 round_trippers.go:423] Response Headers:
I0208 21:40:17.641645    5803 round_trippers.go:426]     Content-Type: application/json
I0208 21:40:17.641647    5803 round_trippers.go:426]     Content-Length: 530
I0208 21:40:17.641649    5803 round_trippers.go:426]     Date: Wed, 08 Feb 2017 21:40:17 GMT
I0208 21:40:17.641667    5803 request.go:991] Response Body: {"apiVersion":"unit-test.test.com/v1","key":"value","kind":"Widget","metadata":{"name":"widget","namespace":"test","selfLink":"/apis/unit-test.test.com/v1/namespaces/test/wi
dgets/widget","uid":"2c81844a-ee47-11e6-a0f6-0800275ae9e7","resourceVersion":"117","creationTimestamp":"2017-02-08T21:40:13Z","annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"unit-test.test.com/v1\"
,\"key\":\"value\",\"kind\":\"Widget\",\"metadata\":{\"annotations\":{},\"name\":\"widget\",\"namespace\":\"test\"}}\n"}}}
dget "widget" configured
```

The unit test adds the kubectl.kubernetes.io/last-applied-configuration to the annotations to the PATCH request.
Maybe you have any idea what may be the reason ?